agent/src/main/java/com/teamscale/jacoco/agent/AgentBase.java

```diff
@@ -45,18 +60,65 @@ public AgentBase(AgentOptions options) throws IllegalStateException {
 
 		logger.info("Starting JaCoCo agent for process {} with options: {}",
 				ManagementFactory.getRuntimeMXBean().getName(), getOptionsObjectToLog());
+		retryUnsuccessfulUploads(options);
 		if (options.getHttpServerPort() != null) {
 			try {
 				initServer();
 			} catch (Exception e) {
-				throw new IllegalStateException(
-						"Control server not started.", e);
+				throw new IllegalStateException("Control server not started.", e);
 			}
 		}
 	}
 
 	/**
-	 * Lazily generated string representation of the command line arguments to print to the log.
+	 * If we have coverage that was leftover because of previously unsuccessful
+	 * coverage uploads, we retry to upload them again with the same configuration
+	 * as in the previous try.
+	 *
+	 * This method expects that the .properties file of the leftover coverage is
+	 * constructed in a specific order.
+	 * 
+	 */
+	private void retryUnsuccessfulUploads(AgentOptions options) {
+		if (options.getOutputDirectory() == null) {
+			return;
+		}
+		List<File> reuploadCandidates = FileSystemUtils.listFilesRecursively(
+				options.getOutputDirectory().getParent().toFile(),
+				filepath -> filepath.getName().contains(RETRY_UPLOAD_FILE_SUFFIX));
+		for (File file : reuploadCandidates) {
+			try {
+				logger.info("Retrying previously unsuccessful coverage upload for file {}.", file);
+				String[] config = FileSystemUtils.readFile(file).split("\n");
```

Comment:
Why not use java `Properties` for writing and parsing the file? This would get rid of the restriction to not change the order of entries and would also work out of the box with line breaks.

Reply:
Would you add constants for the properties then?

Reply:
Sounds like a good idea. You could even group them into an interface.

Reply:
I thought about adding a ETeamscaleServerProperties Enum

Reply:
I don't mind